### PR TITLE
#issue-83 Added UI component Grid generation with data provider generation

### PR DESCRIPTION
**Pull request by serhiyzhovnir**
**Description** (*)
This PR adds the action for generating a simple UI component listing (grid) configuration file with the data provider class.
Screenshot 2020-06-26 at 09 25 35

For the Listing UI component, we can configure the name, identifier field, target area, and ACL (only for the adminhtml area).
Also, we can configure the toolbar features like bookmarks, columns controls, full-text search, listing filters, listing paging.

For data provider, we have two options:
* Custom data provider - it's just a class that inherits the `Magento\Framework\View\Element\UiComponent\DataProvider\DataProvider` class and a developer will be able to reimplement the search criteria processing by himself. (usually for some entities the repository getList may be used or a service)
Screenshot 2020-06-26 at 09 24 48
Screenshot 2020-06-26 at 09 25 01
Screenshot 2020-06-26 at 09 25 09

* Collection data provider - it's the already implemented data provider that processes the search criteria with the collection processor with the selected collection (there is a select field with search where collections as options. Collection is a class that extends the Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection class). This option is mostly could be used for easy models without extension attributes and EAV.
Screenshot 2020-06-26 at 09 23 06
Screenshot 2020-06-26 at 09 23 50
Screenshot 2020-06-26 at 09 24 07

**Fixed Issues (if relevant)**

1. magento/magento2-phpstorm-plugin#83: Action/Code generation. UI component grid generation

**Contribution checklist** (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with integration/functional tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: testData/actions/generation/generator/UiComponentGridDataProviderGenerator/generateCollectionDataProvider/GridDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        RequestInterface $request,
+        FilterBuilder $filterBuilder,
+        CollectionFactory $collectionFactory,
+        CollectionProcessorInterface $collectionProcessor,
+        SearchResultInterfaceFactory $searchResultFactory,
+        array $meta = [],
+        array $data = []
+    )
+    {
```
Comment (eduard13) [inline testData/actions/generation/generator/UiComponentGridDataProviderGenerator/generateCollectionDataProvider/GridDataProvider.php:63]:
```suggestion
    ) {
```

File: testData/actions/generation/generator/UiComponentGridDataProviderGenerator/generateCollectionDataProvider/GridDataProvider.php [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        return $searchResults;
+    }
+
+    /**
+     * Search result to output
+     *
+     * @param SearchResultInterface $searchResult
+     *
+     * @return array|array[]
```
Comment (eduard13) [inline testData/actions/generation/generator/UiComponentGridDataProviderGenerator/generateCollectionDataProvider/GridDataProvider.php:107]:
```suggestion
     * @return array
```

File: testData/actions/generation/generator/UiComponentGridXmlGenerator/generateUiGridForBaseArea/custom_entity_grid.xml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    <columns name="custom_entity_grid_columns">
+        <column name="entity_id">
+            <settings>
+                <filter>textRange</filter>
+                <label translate="true">ID</label>
+                <sorting>desc</sorting>
+            </settings>
+        </column>
+    </columns>
+</listing>
```
Comment (eduard13) [inline testData/actions/generation/generator/UiComponentGridXmlGenerator/generateUiGridForBaseArea/custom_entity_grid.xml:40]:
```suggestion
</listing>

```

File: testData/actions/generation/generator/UiComponentGridXmlGenerator/generateUiGridWithoutToolbar/custom_entity_grid.xml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    <columns name="custom_entity_grid_columns">
+        <column name="entity_id">
+            <settings>
+                <filter>textRange</filter>
+                <label translate="true">ID</label>
+                <sorting>desc</sorting>
+            </settings>
+        </column>
+    </columns>
+</listing>
```
Comment (eduard13) [inline testData/actions/generation/generator/UiComponentGridXmlGenerator/generateUiGridWithoutToolbar/custom_entity_grid.xml:47]:
```suggestion
</listing>

```

**Comment by eduard13** [comment]
#### MAINTAINER REWARD CONCLUSION
Awarding this PR with `Complex`, `Advanced` and `Test Coverage` for its complexity to generate the Ui Listings using the best coding practice, which is the recommended way of building admin grids. Moreover, it provides different ways of DataProvider implementations.

File: testData/actions/generation/generator/UiComponentGridDataProviderGenerator/generateCollectionDataProvider/GridDataProvider.php (hunk 1 above)
Comment (VitaliyBoyko) [reply testData/actions/generation/generator/UiComponentGridDataProviderGenerator/generateCollectionDataProvider/GridDataProvider.php:63]: Let's leave it, it caused by build-in formatted, not related to the plugin

**Review by eduard13 — APPROVED** [review]
Thank you @VitaliyBoyko for helping with it.

**Pull request merged**